flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SlotSharingManager.java

```diff
@@ -0,0 +1,722 @@
+	private final AllocatedSlotActions allocatedSlotActions;
+
+	// owner of the slots to which to return them when they are released from the outside
+	private final SlotOwner slotOwner;
+
+	private final Map<SlotRequestId, TaskSlot> allTaskSlots;
+
+	// Root nodes which have not been completed because the allocated slot is still pending
+	private final Map<SlotRequestId, MultiTaskSlot> unresolvedRootSlots;
+
+	// Root nodes which have been completed (the underlying allocated slot has been assigned)
+	private final Map<TaskManagerLocation, Set<MultiTaskSlot>> resolvedRootSlots;
+
+	// Internal class to iterate over all resolved root slots
+	private ResolvedRootSlotValues resolvedMultiTaskSlotValues;
+
+	public SlotSharingManager(
+			SlotSharingGroupId slotSharingGroupId,
+			AllocatedSlotActions allocatedSlotActions,
+			SlotOwner slotOwner) {
+		this.slotSharingGroupId = Preconditions.checkNotNull(slotSharingGroupId);
+		this.allocatedSlotActions = Preconditions.checkNotNull(allocatedSlotActions);
+		this.slotOwner = Preconditions.checkNotNull(slotOwner);
+
+		allTaskSlots = new HashMap<>(16);
+		unresolvedRootSlots = new HashMap<>(16);
+		resolvedRootSlots = new HashMap<>(16);
+
+		resolvedMultiTaskSlotValues = null;
+	}
+
+	public boolean isEmpty() {
+		return allTaskSlots.isEmpty();
+	}
+
+	public boolean contains(SlotRequestId slotRequestId) {
+		return allTaskSlots.containsKey(slotRequestId);
+	}
+
+	@Nullable
+	public TaskSlot getTaskSlot(SlotRequestId slotRequestId) {
+		return allTaskSlots.get(slotRequestId);
+	}
+
+	/**
+	 * Creates a new root slot with the given {@link SlotRequestId}, {@link SlotContext} future and
+	 * the {@link SlotRequestId} of the allocated slot.
+	 *
+	 * @param slotRequestId of the root slot
+	 * @param slotContextFuture with which we create the root slot
+	 * @param allocatedSlotRequestId slot request id of the underlying allocated slot which can be used
+	 *                               to cancel the pending slot request or release the allocated slot
+	 * @return New root slot
+	 */
+	public MultiTaskSlot createRootSlot(
+			SlotRequestId slotRequestId,
+			CompletableFuture<SlotContext> slotContextFuture,
+			SlotRequestId allocatedSlotRequestId) {
+		final MultiTaskSlot rootMultiTaskSlot = new MultiTaskSlot(
+			slotRequestId,
+			slotContextFuture,
+			allocatedSlotRequestId);
+
+		allTaskSlots.put(slotRequestId, rootMultiTaskSlot);
+		unresolvedRootSlots.put(slotRequestId, rootMultiTaskSlot);
+
+		// add the root node to the set of resolved root nodes once the SlotContext future has been completed
+		// and we know the slot's TaskManagerLocation
+		slotContextFuture.whenComplete(
+			(SlotContext slotContext, Throwable throwable) -> {
+				if (slotContext != null) {
+					final MultiTaskSlot resolvedRootNode = unresolvedRootSlots.remove(slotRequestId);
+
+					if (resolvedRootNode != null) {
+						final Set<MultiTaskSlot> innerCollection = resolvedRootSlots.computeIfAbsent(
```

Comment:
In theory the signature of this method allows concurrent modifications on `resolvedRootSlots`. Maybe a comment why this cannot happen would be nice; perhaps with`@GuardedBy`.

Reply:
True. I'll add the necessary synchronization because it's not guaranteed that the future completion won't change in the future.